### PR TITLE
Optimize Distance Functions - [MOD-5434]

### PR DESCRIPTION
**Pull request by GuyAv46**
**Describe the changes in the pull request**

Refactor all distance functions to be templated by the residual of the dimension, so that each dimension gets a tailor-made implementation with the fewest condition checks possible and at the same time, having less code (one function for each architecture) to maintain.

The new implementation for each architecture uses mask loading of the given arch and therefore we don't have to handle "tails" of non-512-bit-multiple vectors. we handle the residual part first, so the mask loading can safely load some more data if the implementation requires it, and with the 512-bit chunks after. We assume we have enough owned data for the mask loading, so we give up optimizing dimensions smaller than 512-bit (16 for floats, 8 for doubles). This is fine because there is no too-big hit, and there are no real-life scenarios of small dimensions. We can implement spatial optimizations for small dimensions in the future.

**Main objects this PR modified**
1. All distance functions.

**Mark if applicable**

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

### CODE REVIEW

File: src/VecSim/spaces/IP_space.cpp [hunk 1]
```diff
@@ -6,121 +6,85 @@
 
 #include "VecSim/spaces/IP_space.h"
 #include "VecSim/spaces/IP/IP.h"
-#include "VecSim/spaces/IP/IP_AVX512DQ.h"
+#if defined(__x86_64__)
 #include "VecSim/spaces/IP/IP_AVX512.h"
 #include "VecSim/spaces/IP/IP_AVX.h"
 #include "VecSim/spaces/IP/IP_SSE.h"
```
Comment (meiravgri) [inline src/VecSim/spaces/IP_space.cpp:12]: Are we sure it is dafe to include higher optimization in older platforms? (AVX in SSE for example)

**Review by meiravgri — COMMENTED** [review]
looks very good and clean ! Well done
Please verify we still need all the compilation & hardware flags for DQ

File: src/VecSim/spaces/CMakeLists.txt [hunk 2]
```diff
@@ -65,54 +65,3 @@ if(NOT APPLE)
 elseif(CMAKE_HOST_SYSTEM_PROCESSOR STREQUAL "x86_64")
 	target_link_libraries(VectorSimilaritySpaces cpu_features)
 endif()
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/spaces/CMakeLists.txt:67]: We can, but we probably don't have to. As far as I could see there was no effect on the assembly output, so I will remove it and any reference to it from the project

File: tests/unit/test_spaces.cpp [hunk 3]
```diff
@@ -98,303 +90,149 @@ TEST_F(SpacesTest, double_ip_no_optimization_func_test) {
...
-    }
+TEST_F(SpacesTest, smallDimChooser) {
+    size_t dim = 5;
+
+    // Verify that small dimensions gets the no optimization function
+    ASSERT_EQ(L2_FP32_GetDistFunc(dim, ARCH_OPT_AVX512_F), FP32_L2Sqr);
+    ASSERT_EQ(L2_FP64_GetDistFunc(dim, ARCH_OPT_AVX512_F), FP64_L2Sqr);
+    ASSERT_EQ(IP_FP32_GetDistFunc(dim, ARCH_OPT_AVX512_F), FP32_InnerProduct);
+    ASSERT_EQ(IP_FP64_GetDistFunc(dim, ARCH_OPT_AVX512_F), FP64_InnerProduct);
 }
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply tests/unit/test_spaces.cpp:103]: Unless we make some optimization for dim<8/16

File: src/VecSim/spaces/implementation_chooser_cleanup.h [hunk 4]
```diff
@@ -0,0 +1,17 @@
+/*
+ *Copyright Redis Ltd. 2021 - present
+ *Licensed under your choice of the Redis Source Available License 2.0 (RSALv2) or
+ *the Server Side Public License v1 (SSPLv1).
+ */
+
+#pragma once
+
+/*
+ * Include this file after done using the implementation chooser.
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/spaces/implementation_chooser_cleanup.h:10]: probably nothing unless more X definitions will be presented

File: src/VecSim/spaces/implementation_chooser.h [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+ */
+
+#pragma once
+
+/*
+ * This file contains macros magic to choose the implementation of a function based on the
+ * dimension's remainder. It is used to collapse large and repetitive switch statements that are
+ * used to choose and define the templated values of the implementation of the distance functions.
+ * We assume that we are dealing with 512-bit blocks, so we define a chunk size of 16 for 32-bit
+ * elements, and a chunk size of 8 for 64-bit elements. The main macro is CHOOSE_IMPLEMENTATION, and
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/spaces/implementation_chooser.h:14]: this was the way the functions were implemented in the first place. we can consider changing that but today it simply means X2 more condition checks for AVX and X4 for SSE. also it makes choosing the implementation simpler

File: src/VecSim/spaces/implementation_chooser.h [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+        __ret_dist_func = func<(N)>;                                                               \
+        break;
+
+// Macro for 4 cases. Used to collapse the switch statement. For a given N, expands to 4 X macros
+// of 4N, 4N+1, 4N+2, 4N+3.
+#define C4(X, func, N) X(4 * N, func) X(4 * N + 1, func) X(4 * N + 2, func) X(4 * N + 3, func)
+
+// Macros for 8 and 16 cases. Used to collapse the switch statement. Expands into 0-15 or 0-7 cases.
+#define CASES16(X, func) C4(X, func, 0) C4(X, func, 1) C4(X, func, 2) C4(X, func, 3)
+#define CASES8(X, func)  C4(X, func, 0) C4(X, func, 1)
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/spaces/implementation_chooser.h:30]: Right. This was done for readability (so each definition will fit one line). I can change that

File: src/VecSim/spaces/CMakeLists.txt [hunk 7]
```diff
@@ -65,54 +59,3 @@ if(NOT APPLE)
...
-
-if(CMAKE_HOST_SYSTEM_PROCESSOR MATCHES "(x86_64)|(AMD64|amd64)|(^i.86$)")
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/spaces/CMakeLists.txt:69]: as all the functions are now templated and written in H files, we no longer have files that are compiled only sometimes

File: src/VecSim/spaces/AVX_utils.h [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+ *Copyright Redis Ltd. 2021 - present
+ *Licensed under your choice of the Redis Source Available License 2.0 (RSALv2) or
+ *the Server Side Public License v1 (SSPLv1).
+ */
+
+#pragma once
+#include "space_includes.h"
+
+template <__mmask8 mask> // (2^n)-1, where n is in 1..7 (1, 4, ..., 127)
+static inline __m256 my_mm256_maskz_loadu_ps(const float *p) {
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/spaces/AVX_utils.h:11]: I can make them not templated. Probably won't affect the performance

**Review by DvirDukhan — APPROVED** [review]

**Pull request merged**